File: clients/windows/Mailcal/Services/CrashLog.cs
// What the diagnostic log says when something is about to kill the app (docs/logging.md → "A
// crash says so on the way out"). Without it a crash is indistinguishable in the file from a clean
// exit: the log simply stops, with nothing wrong on the last line.
//
// Added because one did. A COMException out of NavigationView.IsPaneOpen took the app down as a
// stowed exception inside Microsoft.UI.Xaml.dll, and the only record of it anywhere on the machine
// was a Windows Error Reporting entry naming an offset in combase.dll.
//
// WinUI-free on purpose, so the line composition is reachable from Mailcal.Tests (which cannot
// link WinUI types). The one handler that needs the Application object, WinUI's own
// UnhandledException, for XAML-thread faults, is wired in App.xaml.cs and formats its line
// through Record here, so all four shapes say the same thing.

using System.Threading.Tasks;

namespace Allodia.Mailcal.Services;

/// <summary>Writes the stack of anything about to kill the process into the diagnostic log.</summary>
internal static class CrashLog
{
    /// <summary>
    /// Wires the two runtime-wide handlers. Called from <c>Program.Main</c> immediately after
    /// <c>Log.Init</c>, the ordering is the point, since <c>Log.Write</c> silently drops
    /// everything until the sink has a path.
    /// </summary>
    /// <remarks>
    /// Neither handler marks the fault handled: the process is meant to still fail, and this only
    /// makes it say why on the way out.
    /// </remarks>
    internal static void WatchProcess()
    {
        // The CLR's own: everything off the XAML thread, including a background thread that throws.
        AppDomain.CurrentDomain.UnhandledException += (_, e) =>
            Log.Error(Record(e.IsTerminating ? "on a terminating thread" : "on a thread", e.ExceptionObject));

        // A Task nobody awaited, whose exception the finalizer would otherwise swallow whole. This
        // one is NOT a crash, since .NET 4.5 the process survives it, which is exactly why it is
        // worth a line: an `async void` handler or a fire-and-forget `_ =` call that faults leaves
        // no other trace anywhere. SetObserved is deliberately not called; not calling it keeps the
        // framework's default behaviour rather than changing it.
        TaskScheduler.UnobservedTaskException += (_, e) =>
            Log.Error(Record("in a task nobody awaited, which did not stop the app", e.Exception));
    }

    /// <summary>
    /// Arms the core's native-fault handler over the same file.
    /// </summary>
    /// <remarks>
    /// An access violation inside the cdylib reaches none of the handlers above: the CLR tears the
    /// process down without raising <c>AppDomain.UnhandledException</c>, so the log stops mid-line
    /// and Windows Error Reporting holds the only record. The core installs a vectored exception
    /// handler for it, which observes and never handles, the process still fails exactly as it
    /// would have.
    /// <para>
    /// Kept out of <see cref="WatchProcess"/> because this is the one call here that crosses the
    /// FFI, and <c>Mailcal.Tests</c> links no cdylib; keeping them apart is what lets the record
    /// composition above stay unit-tested. A missing library is swallowed for the reason every
    /// other log failure is: diagnostics must never be what stops the app from starting, and a
    /// cdylib that is genuinely absent takes the app down at its first real call regardless.
    /// </para>
    /// </remarks>
    internal static void WatchForNativeFaults(string logPath)
    {
        try
        {
            uniffi.mailcal_bindings.MailcalBindingsMethods.WatchForNativeFaults(logPath);
        }
        catch (Exception e)
        {
            Log.Warn($"the native-fault handler could not be armed: {e.GetType().Name}");
        }
    }

    /// <summary>The line one fault writes.</summary>
    /// <remarks>
    /// <c>unhandled</c> is the word every platform's crash line carries, one string support greps
    /// across four clients, and a .NET exception's <c>ToString()</c> already opens with
    /// <c>Type: Message</c> and continues into the frames, so the phrase says only what the type
    /// cannot: which part of the runtime the fault came out of.
    /// <para>
    /// The fault arrives as <see cref="object"/> because <c>AppDomain.UnhandledException</c> hands
    /// over an <c>ExceptionObject</c>, which is only an <c>Exception</c> by convention.
    /// </para>
    /// </remarks>
    internal static string Record(string where, object? fault) => $"unhandled {where}: {fault}";
}

File: clients/android/app/src/main/java/eu/allodia/mailcal/CrashLog.kt
// The uncaught-exception handler: what the diagnostic log says when the JVM half of the app dies.
//
// Without it a crash is indistinguishable in the file from a clean exit, the log simply stops,
// with nothing wrong on the last line (docs/logging.md → "A crash says so on the way out"). This
// covers every JVM thread in the process, WorkManager workers included.
//
// It does NOT cover the two native cases, which have handlers of their own in the core: a Rust
// panic writes through the shared hook (crates/mailcal-bindings/src/crash.rs) and a SIGSEGV or an
// abort inside the cdylib through the native-fault handler `watchForNativeFaults` arms below.
package eu.allodia.mailcal

object CrashLog {
    // Arms the handler. Called from MailcalApplication.onCreate immediately after FileLog.init, so
    // it is in place on both process entry points, a user launch and a cold WorkManager wake.
    fun watchForCrashes() {
        Thread.setDefaultUncaughtExceptionHandler(
            handler(Thread.getDefaultUncaughtExceptionHandler()),
        )
    }

    // The handler itself, over the handler it replaces.
    //
    // **Chaining is not politeness.** Android's own default handler is what shows the crash dialog,
    // reports to Play Console, and kills the process; swallowing it leaves a dead app that never
    // says it died. So the record is written first and `previous` runs after, FileLog.append is
    // synchronous under its lock, so the line is on disk before the process goes.
    //
    // Taking `previous` as a parameter rather than reading it here keeps the global handler out of
    // the test JVM: the wiring above is the only thing that mutates process state.
    internal fun handler(previous: Thread.UncaughtExceptionHandler?): Thread.UncaughtExceptionHandler =
        Thread.UncaughtExceptionHandler { thread, throwable ->
            FileLog.append("ERROR", "crash", record(thread.name, throwable))
            previous?.uncaughtException(thread, throwable)
        }

    // The record. `unhandled` is the word every platform's crash line carries, one string support
    // greps across four clients, and the stack's own first line already reads `<type>: <message>`,
    // so the headline names the thread and the frames say the rest without repeating it.
    internal fun record(thread: String, throwable: Throwable): String =
        "unhandled on $thread: ${throwable.stackTraceToString().trimEnd()}"

    // Arms the core's native-fault handler over the same file. A SIGSEGV or an abort inside the
    // cdylib is neither a JVM throwable nor a Rust panic, so the handler above never sees it and
    // the log just stops; this writes one banner naming the signal and the faulting address. The
    // handler chains to Android's own, so the tombstone and the Play Console report are unchanged.
    //
    // It takes the path rather than being folded into `watchForCrashes` because this is the one
    // call here that crosses the FFI, and `:app:test` loads no cdylib, keeping them apart is what
    // lets the handler above stay unit-tested. A missing library is swallowed for the same reason
    // every other log failure is: diagnostics must never be what stops the app from starting, and
    // a cdylib that is genuinely absent takes the app down at its first real call regardless.
    fun watchForNativeFaults(logPath: String) {
        try {
            uniffi.mailcal_bindings.watchForNativeFaults(logPath)
        } catch (_: Throwable) {
        }
    }
}
